src/Service/IncentiveService.php

```diff
@@ -0,0 +1,133 @@
+<?php
+
+namespace App\Service;
+
+use Doctrine\ORM\EntityManagerInterface;
+use App\Audit\Log;
+
+class IncentiveService
+{
+    protected $rdrApiService;
+    protected $siteService;
+    protected $userService;
+    protected $em;
+    protected $loggerService;
+
+    public function __construct(
+        RdrApiService $rdrApiService,
+        SiteService $siteService,
+        UserService $userService,
+        EntityManagerInterface $em,
+        LoggerService $loggerService
+    ) {
+        $this->rdrApiService = $rdrApiService;
+        $this->siteService = $siteService;
+        $this->userService = $userService;
+        $this->em = $em;
+        $this->loggerService = $loggerService;
+    }
+
+    public function getRdrObject($incentive, $type = 'create')
+    {
+        $obj = new \StdClass();
+        $email = $this->userService->getUser()->getEmail();
+        $now = new \DateTime();
+        if ($type !== 'create') {
+            $obj->incentiveId = $incentive->getRdrid();
+        }
+        if ($type === 'cancel') {
+            $obj->cancelledBy = $email;
+            $obj->cancelledDate = $now;
+            $obj->cancel = true;
+        } else {
+            $obj->createdBy = $email;
+            $obj->site = $this->siteService->getSiteIdWithPrefix();
+            $obj->dateGiven = $incentive->getIncentiveDateGiven();
+            $obj->occurrence = $incentive->getOtherIncentiveOccurrence() ?? $incentive->getIncentiveOccurrence();
+            $obj->incentiveType = $incentive->getOtherIncentiveType() ?: $incentive->getIncentiveType();
+            if ($incentive->getGiftCardType()) {
+                $obj->giftcardType = $incentive->getGiftCardType();
+            }
+            $obj->amount = $incentive->getIncentiveAmount();
+            $obj->notes = $incentive->getNotes();
+        }
+        return $obj;
+    }
+
+    public function createIncentive($participantId, $incentiveForm)
+    {
+        $incentive = $incentiveForm->getData();
+        if ($incentive->getIncentiveAmount() === 'other') {
+            $incentive->setIncentiveAmount($incentiveForm['other_incentive_amount']->getData());
+        }
```

Comment:
Can this logic be implemented as a data mapper in the form?

Reply:
Do you have an example of how to implement that using data mapper?

Reply:
I might have gotten data mappers and data transformers confused.

Reading through the documentation, I _think_ a data transformer makes sense here: https://symfony.com/doc/5.4/form/data_transformers.html

And here's the section in the data mapper documentation about the difference between mappers and transformers: https://symfony.com/doc/5.4/form/data_mappers.html#the-difference-between-data-transformers-and-mappers

I don't have a specific example, but the documentation provides a few. If this doesn't make sense in this case, that's ok. I was just looking for a way to not have to repeat this logic in both the create and amend methods.